manager/controlapi/service.go

```diff
@@ -175,6 +176,49 @@ func validateServiceSpec(spec *api.ServiceSpec) error {
 	return nil
 }
 
+func (s *Server) checkPortConflicts(spec *api.ServiceSpec) error {
+	if spec.Endpoint == nil {
+		return nil
+	}
+
+	var (
+		services []*api.Service
+		err      error
+	)
+
+	s.store.View(func(tx store.ReadTx) {
+		services, err = store.FindServices(tx, store.All)
+	})
```

Comment:
Since `FindServices` happens inside its own transaction, the check and the service create/update are not an atomic set of operations, and it's potentially possible to do something like create two services simultaneously that use the same port.

This may be the correct tradeoff, because holding a lock over the store for the whole time (putting everything inside an `Update` transaction) could potentially block for awhile, since `FindServices` could be expensive. But it's important to note that the current code is a "best effort" solution.